pkg/options/options.go

```diff
@@ -0,0 +1,129 @@
+	ProxyImage string
+}
+
+func NewCoreEtcdOptions() *CoreEtcdOptions {
+	return &CoreEtcdOptions{
+		URL:             "",
+		CAConfigMapName: "etcd-coreserving-ca",
+		CertSecretName:  "etcd-coreserving-cert",
+	}
+}
+
+func NewEtcdProxyControllerOptions() *EtcdProxyControllerOptions {
+	return &EtcdProxyControllerOptions{
+		CoreEtcd:            NewCoreEtcdOptions(),
+		ControllerNamespace: "kube-apiserver-storage",
+		KubeconfigPath:      "",
+		ProxyImage:          "quay.io/coreos/etcd:v3.2.18",
+	}
+}
+
+func (e *EtcdProxyControllerOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVarP(&e.CoreEtcd.URL, "etcd-core-url", "u", e.CoreEtcd.URL, "The address of the core etcd server.")
+	fs.StringVar(&e.CoreEtcd.CAConfigMapName, "etcd-core-ca-configmap", e.CoreEtcd.CAConfigMapName, "The name of the ConfigMap where CA is stored.")
+	fs.StringVar(&e.CoreEtcd.CertSecretName, "etcd-core-certs-secret", e.CoreEtcd.CertSecretName, "The name of the Secret where client certificates are stored.")
+
+	fs.StringVarP(&e.ControllerNamespace, "namespace", "n", e.ControllerNamespace, "Name of the namespace where controller is deployed.")
+	fs.StringVarP(&e.KubeconfigPath, "kubeconfig", "k", e.KubeconfigPath, "Path to kubeconfig (required only if running out-cluster).")
+	fs.StringVar(&e.ProxyImage, "etcd-proxy-image", e.ProxyImage, "The image to be used for creating etcd proxy pods.")
+}
+
+func (e *EtcdProxyControllerOptions) ApplyTo(c *etcdproxy.EtcdProxyControllerConfig) error {
+	var err error
+
+	c.CoreEtcd = &etcdproxy.CoreEtcdConfig{}
+	c.CoreEtcd.URL = e.CoreEtcd.URL
+	c.CoreEtcd.CAConfigMapName = e.CoreEtcd.CAConfigMapName
+	c.CoreEtcd.CertSecretName = e.CoreEtcd.CertSecretName
+
+	c.ControllerNamespace = e.ControllerNamespace
+	c.ProxyImage = e.ProxyImage
+
+	c.Kubeconfig, err = clientcmd.BuildConfigFromFlags("", e.KubeconfigPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EtcdProxyControllerOptions) Validate() error {
+	errors := []error{}
+
+	errors = append(errors, e.CoreEtcd.Validate())
+
+	if e.ControllerNamespace == "" {
+		errors = append(errors, fmt.Errorf("controller namespace name empty"))
+	}
+
+	if e.KubeconfigPath == "" {
+		errors = append(errors, fmt.Errorf("kubeconfig path empty"))
+	}
+
+	if e.ProxyImage == "" {
+		errors = append(errors, fmt.Errorf("etcd proxy image name empty"))
+	}
+
+	return utilerrors.NewAggregate(errors)
+}
+
+func (c *CoreEtcdOptions) Validate() error {
+	errors := []error{}
+
+	if c.URL == "" {
+		errors = append(errors, fmt.Errorf("core etcd url empty"))
+	}
+
+	if c.CAConfigMapName == "" {
+		errors = append(errors, fmt.Errorf("core etcd ca certificates configmap name empty"))
+	}
+
+	if c.CertSecretName == "" {
+		errors = append(errors, fmt.Errorf("core etcd certificates secret name empty"))
+	}
+
+	return utilerrors.NewAggregate(errors)
+}
+
+func (e *EtcdProxyControllerOptions) Config() (*etcdproxy.EtcdProxyControllerConfig, error) {
```

Comment:
nit: I prefer the ApplyTo construct so that I can easily have multiple options applied to different aspects of the same config type.